src/kdc/kdc_util.c

```diff
@@ -265,15 +237,6 @@ kdc_process_tgs_req(kdc_realm_t *kdc_active_realm,
         goto cleanup_authenticator;
     }
 
-    /* make sure the client is of proper lineage (see above) */
-    if (foreign_server && !has_s4u2self_padata(request->padata) &&
-        is_local_principal(kdc_active_realm, ticket->enc_part2->client)) {
-        /* someone in a foreign realm claiming to be local */
-        krb5_klog_syslog(LOG_INFO, _("PROCESS_TGS: failed lineage check"));
-        retval = KRB5KDC_ERR_POLICY;
-        goto cleanup_authenticator;
-    }
```

Comment:
I'm afraid I don't see why it's okay to remove this check; do you mind clarifying?

Reply:
This check is moved to do_tgs_req.c line 297, after the kdc_process_s4u2self_req() call, where we know in a more natural way whether the request is an S4U2Self request and where we have the canonicalized server principal.  I will call that out in the commit message.

Where this check should live is an open question.  I don't think it really ever belonged in kdc_process_tgs_req() since it isn't about decoding the header authenticator.  (That function also contains an FX_ARMOR check which I think belongs elsewhere.)  Putting it in do_tgs_req.c makes a long function longer, which I don't like, but validate_tgs_request() doesn't have quite the right inputs (whether this is a cross-realm request, whether this is an S4U2Self request) and I'm still considering whether there is a natural way to change that.  I would prefer not to hold up this PR for that analysis, instead perhaps doing some refactoring work after the fact.

Reply:
Thanks, I think I understand now.  I agree it's not worth holding up the PR.